tools/codegen/dest/native_functions.py

```diff
@@ -53,3 +58,138 @@ def compute_native_function_declaration(
+        // rely on other behavior in the lowering or copy process to make this correct.
+        // (1) evaluate design goal: to always pick the IR's dtype in one place (here)
+        // (2) rationalize this with Google's design, it may be a problem
+        // (3) evaluate perf impact: make sure we're not actually doing casts becuase of this override
+        _out_dtype));
+}};
+"""
+
+
+def compute_lazy_native_function_definition(
+        g: Union[NativeFunctionsGroup, NativeFunction],
+        backend_index: BackendIndex,
+        codegen: List[OperatorName],
+        class_method_name: str,
+) -> List[str]:
+
+    metadata = backend_index.get_kernel(g)
+    if isinstance(g, NativeFunctionsGroup):
+        if metadata is not None and metadata.structured:
+            raise AssertionError("Structured lazy functions are not implemented yet.")
+        else:
+            return list(mapMaybe(lambda f: gen_unstructured_lazy_definition(f, backend_index, codegen, class_method_name), g.functions()))
+    else:
+        x = gen_unstructured_lazy_definition(g, backend_index, codegen, class_method_name)
+        return [] if x is None else [x]
+
+
+def gen_lazy_shape_dtype_decl(f: NativeFunction, backend_index: BackendIndex, codegen: List[OperatorName]) -> Optional[str]:
+    sig = kernel_signature(f, backend_index)
+    metadata = backend_index.get_kernel(f)
+    if f.func.name not in codegen:
+        return None
+    if metadata is None:
+        return None
+    if "legacy::" in metadata.kernel:
+        return None
+
+    # Lazy IR stuff
+    schema = update_schema_for_lazy_ir(f.func)
+    all_types, value_types, scalar_types = separate_value_scalar_types(schema)
+    lazy_tensor_decls_str = lazy_tensor_decls(value_types)
+    node_ctor_input_str = node_ctor_inputs(value_types, scalar_types)
+
+    # Only generate shape/dtype fn for non-structured kernels,
+    # since we just use the meta function for structured kernels
+    if not f.structured and f.structured_delegate is None:
+        shapeFn = DispatcherSignature.from_schema(FunctionSchema(f.func.name,
+                                                                 f.func.arguments,
+                                                                 returns=(Return(name=None,
+                                                                                 type=intArrayT,
+                                                                                 annotation=None),)),
+                                                  prefix='compute_shape_')
+        dtypeFn = DispatcherSignature.from_schema(FunctionSchema(f.func.name,
+                                                                 f.func.arguments,
+                                                                 returns=(Return(name=None,
+                                                                                 type=scalarTypeT,
```

Comment:
@bdhirsh I had trouble here.  Intended to create new DispatcherSignature objects with updated Returntypes, and use those to generate the code below (the function signature declarations).

The problem was if i called shapeFn.decl() it complained that scalarTypeT was invalid as a return type.  I didn't really understand the assumptions/logic of ReturnType, so I just gave up and wrote the code below.  

If you think there is a quick or longer-term-worthwhile fix for my above problem, i'll leave this code in place; if not, I can delete these and just use the code below.